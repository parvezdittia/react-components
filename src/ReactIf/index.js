let ReactIf;
let ReactElseIf;
ReactIf = ReactElseIf = (props) => {
	let output = null;

	if (props.when) {
		output = props.children;
	}

	return output;
};

function ReactElse(props) {
	let output = null;

	if (props.children) {
		output = props.children;
	}

	return output;
}

function ReactChain(props) {
	let links = props.children;
	let output = null;

	if (!links.length) {
		console.error('ReactChain expect conditional component');
		return output;
	}

	for (let i = 0; i < links.length; i++) {
		try {
			if (
				links[i].type.name !== 'ReactIf' &&
				links[i].type.name !== 'ReactElseIf' &&
				links[i].type.name !== 'ReactElse'
			) {
				throw new Error('Not a valid conditional component');
			}
		} catch (error) {
			console.error(error);
			return output;
		}
	}

	for (let i = 0; i < links.length; i++) {
		if (links[i].props.when) {
			output = links[i];
			break;
		}
	}

	if (output === null && links[links.length - 1].type.name === 'ReactElse') {
		output = links[links.length - 1];
	}

	return output;
}

export { ReactChain, ReactElseIf, ReactElse };
export default ReactIf;
