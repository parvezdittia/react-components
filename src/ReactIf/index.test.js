import React from 'react';
import renderer from 'react-test-renderer';
import ReactIf from './index';
import { describe, expect, it } from '@jest/globals';

describe.each([
	true,
	1,
	-1,
	'abc',
	{},
	{ a: 1 },
	[],
	['a'],
	(() => {
		return true;
	})(),
])('Render "when" boolean value is true', (condition) => {
	const placeHolderComponent = <span>Hello World</span>;
	it(`Renders child component if condition is ${condition}`, () => {
		const testComponent = (
			<ReactIf when={condition}>{placeHolderComponent}</ReactIf>
		);
		const tree = renderer.create(testComponent).toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe.each([undefined, false, 0, '', null])(
	'Render "when" boolean value is false',
	(condition) => {
		const placeHolderComponent = <span>Hello World</span>;
		it(`Renders null if condition is ${condition}`, () => {
			const testComponent = (
				<ReactIf when={condition}>{placeHolderComponent}</ReactIf>
			);
			const tree = renderer.create(testComponent).toJSON();
			expect(tree).toMatchSnapshot();
		});
	}
);

describe.each([
	[false, false, null],
	[false, true, null],
	[true, false, null],
	[true, true, 'Child Component'],
])('Render nested ReactIfs', (outerCondition, innerCondition, output) => {
	const placeHolderComponent = <span>Hello World</span>;
	const innerReactIf = (
		<ReactIf when={innerCondition}>{placeHolderComponent}</ReactIf>
	);
	const testComponent = <ReactIf when={outerCondition}>{innerReactIf}</ReactIf>;
	it(`OuterCondition: ${outerCondition} and InnerCondition: ${innerCondition} = Output ${output}`, () => {
		const tree = renderer.create(testComponent).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
