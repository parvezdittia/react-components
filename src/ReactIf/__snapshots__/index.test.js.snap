// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render "when" boolean value is false Renders null if condition is  1`] = `null`;

exports[`Render "when" boolean value is false Renders null if condition is 0 1`] = `null`;

exports[`Render "when" boolean value is false Renders null if condition is false 1`] = `null`;

exports[`Render "when" boolean value is false Renders null if condition is null 1`] = `null`;

exports[`Render "when" boolean value is false Renders null if condition is undefined 1`] = `null`;

exports[`Render "when" boolean value is true Renders child component if condition is  1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is [object Object] 1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is [object Object] 2`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is -1 1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is 1 1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is a 1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is abc 1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is true 1`] = `
<span>
  Hello World
</span>
`;

exports[`Render "when" boolean value is true Renders child component if condition is true 2`] = `
<span>
  Hello World
</span>
`;

exports[`Render nested ReactIfs OuterCondition: false and InnerCondition: false = Output null 1`] = `null`;

exports[`Render nested ReactIfs OuterCondition: false and InnerCondition: true = Output null 1`] = `null`;

exports[`Render nested ReactIfs OuterCondition: true and InnerCondition: false = Output null 1`] = `null`;

exports[`Render nested ReactIfs OuterCondition: true and InnerCondition: true = Output Child Component 1`] = `
<span>
  Hello World
</span>
`;
